pyartifactory/objects/artifact.py

```diff
@@ -0,0 +1,322 @@
+from __future__ import annotations
+
+import logging
+import os
+from collections.abc import Iterator
+from pathlib import Path
+from typing import Dict, List, Optional, Union
+
+import requests
+from pydantic import TypeAdapter
+
+from pyartifactory.exception import ArtifactNotFoundError, ArtifactoryError, BadPropertiesError, PropertyNotFoundError
+from pyartifactory.models.artifact import (
+    ArtifactFileInfoResponse,
+    ArtifactFolderInfoResponse,
+    ArtifactInfoResponse,
+    ArtifactListResponse,
+    ArtifactPropertiesResponse,
+    ArtifactStatsResponse,
+)
+from pyartifactory.objects.object import ArtifactoryObject
+
+logger = logging.getLogger("pyartifactory")
+
+
+class ArtifactoryArtifact(ArtifactoryObject):
+    """Models an artifactory artifact."""
+
+    def _walk(self, artifact_path: str, topdown: bool = True) -> Iterator[ArtifactInfoResponse]:
+        """Iterate over artifact (file or directory) recursively.
+
+        :param artifact_path: Path to file or folder in Artifactory
+        :param topdown: True for a top-down (directory first) traversal
+        """
+        info = self.info(artifact_path)
+        if not isinstance(info, ArtifactFolderInfoResponse):
+            yield info
+        else:
+            if topdown:
+                yield info
+            for subdir in (child for child in info.children if child.folder is True):
+                yield from self._walk(artifact_path + subdir.uri, topdown=topdown)
+            for file in (child for child in info.children if child.folder is not True):
+                yield from self._walk(artifact_path + file.uri, topdown=topdown)
+            if not topdown:
+                yield info
+
+    def info(self, artifact_path: Union[Path, str]) -> ArtifactInfoResponse:
+        """
+        Retrieve information about a file or a folder
+
+        See https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-FolderInfo
+        and https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-FileInfo
+
+        :param artifact_path: Path to file or folder in Artifactory
+        """
+
+        if isinstance(artifact_path, str):
+            artifact_path = Path(artifact_path)
+
+        try:
+            response = self._get(f"api/storage/{artifact_path.as_posix()}")
+            return TypeAdapter(ArtifactInfoResponse).validate_python(response.json())  # type: ignore
+        except requests.exceptions.HTTPError as error:
+            if error.response.status_code == 404:
+                logger.error("Artifact %s does not exist", artifact_path)
+                raise ArtifactNotFoundError(f"Artifact {artifact_path} does not exist")
+            raise ArtifactoryError from error
+
+    def deploy(self, local_file_location: Path, artifact_path: Path) -> ArtifactInfoResponse:
```

Comment:
If the `local_file_location` and `artifact_path` are Path and not str any more. The README will need to be modified. Personally, I prefer them to be str

Reply:
It would actually be preferable to have a `Union[Path, str]` if you want to use Path. Also, all other functions that deal with paths must be changed (out of scope of this Merge request IMO)

Reply:
changed it to Union as suggested